components/ws-manager/pkg/manager/manager.go

```diff
@@ -346,6 +361,117 @@ func (m *Manager) StartWorkspace(ctx context.Context, req *api.StartWorkspaceReq
 	return okResponse, nil
 }
 
+func (m *Manager) restoreVolumeSnapshotFromHandle(ctx context.Context, id, handle string) (err error) {
+	span, ctx := tracing.FromContext(ctx, "restoreVolumeSnapshotFromHandle")
+	defer tracing.FinishSpan(span, &err)
+	//log := log.WithField("func", "restoreVolumeSnapshotFromHandle")
+
+	// restore is a two step process
+	// 1. create volume snapshot referencing not yet created volume snapshot content
+	// 2. create volume snapshot content referencing volume snapshot
+	// this creates correct bidirectional binding between those two
+
+	// todo(pavel): figure out if there is a way to find out which snapshot class we need to use here. For now use default class info.
+	var volumeSnapshotClassName string
+	if m.Config.WorkspaceClasses[config.DefaultWorkspaceClass].PVC.SnapshotClass != "" {
+		volumeSnapshotClassName = m.Config.WorkspaceClasses[config.DefaultWorkspaceClass].PVC.SnapshotClass
+	}
+
+	var volumeSnapshotClass volumesnapshotv1.VolumeSnapshotClass
+	err = m.Clientset.Get(ctx, types.NamespacedName{Namespace: "", Name: volumeSnapshotClassName}, &volumeSnapshotClass)
+	if err != nil {
+		return fmt.Errorf("was unable to get volume snapshot class: %v", err)
+	}
+
+	snapshotContentName := "restored-" + id
+	volumeSnapshot := &volumesnapshotv1.VolumeSnapshot{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      id,
+			Namespace: m.Config.Namespace,
+		},
+		Spec: volumesnapshotv1.VolumeSnapshotSpec{
+			Source: volumesnapshotv1.VolumeSnapshotSource{
+				VolumeSnapshotContentName: &snapshotContentName,
+			},
+		},
+	}
+	if volumeSnapshotClassName != "" {
+		volumeSnapshot.Spec.VolumeSnapshotClassName = &volumeSnapshotClassName
+	}
+
+	err = m.Clientset.Create(ctx, volumeSnapshot)
+	if err != nil && !k8serr.IsAlreadyExists(err) {
+		return fmt.Errorf("cannot create volumesnapshot: %v", err)
+	}
+
+	volumeSnapshotContent := &volumesnapshotv1.VolumeSnapshotContent{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      snapshotContentName,
+			Namespace: "", // content is not namespaced
+		},
+		Spec: volumesnapshotv1.VolumeSnapshotContentSpec{
+			VolumeSnapshotRef: corev1.ObjectReference{
+				Kind:      "VolumeSnapshot",
+				Name:      id,
+				Namespace: m.Config.Namespace,
+			},
+			DeletionPolicy: "Delete",
+			Source: volumesnapshotv1.VolumeSnapshotContentSource{
+				SnapshotHandle: &handle,
+			},
+			Driver: volumeSnapshotClass.Driver,
+		},
+	}
+
```

Comment:
Nit: We could consider add VolumeSnapshotClassName to VolumeSnapshotContent as well.
```go
if volumeSnapshotClassName != "" {
	volumeSnapshotContent.Spec.VolumeSnapshotClassName = &volumeSnapshotClassName
}
```

Reply:
Not sure I understood what you meant.

Reply:
Because we add volumeSnapshotClassName to VolumeSnapshot object on lines 397-399.
I'm thinking that we could add volumeSnapshotClassName to VolumeSnapshotContent object as well. (Although if we don't specify the VolumeSnapshotClass, the Kubernetes uses the default one within the cluster).